File: modules/common/common-domain/src/main/scala/com/monadial/waygrid/common/domain/model/resiliency/RetryPolicy.scala
package com.monadial.waygrid.common.domain.model.resiliency

import scala.concurrent.duration.*

/**
 * Represents a general retry policy for any operation (network, IO, routing, etc).
 * Pure, domain-level description — no side effects or runtime semantics.
 */
sealed trait RetryPolicy

sealed trait RetryPolicyWithMaxRetries extends RetryPolicy:
  def maxRetries: Int

sealed trait RetryPolicyWithBaseDelay extends RetryPolicy:
  def base: FiniteDuration

object RetryPolicy:
  case object None extends RetryPolicy

  final case class Linear(
    base: FiniteDuration,
    maxRetries: Int
  ) extends RetryPolicy with RetryPolicyWithBaseDelay with RetryPolicyWithMaxRetries

  final case class Exponential(
    base: FiniteDuration,
    maxRetries: Int
  ) extends RetryPolicy with RetryPolicyWithBaseDelay with RetryPolicyWithMaxRetries

  final case class BoundedExponential(
    base: FiniteDuration,
    cap: FiniteDuration,
    maxRetries: Int
  ) extends RetryPolicy with RetryPolicyWithBaseDelay with RetryPolicyWithMaxRetries

  final case class Fibonacci(
    base: FiniteDuration,
    maxRetries: Int
  ) extends RetryPolicy with RetryPolicyWithBaseDelay with RetryPolicyWithMaxRetries

  final case class Polynomial(
    base: FiniteDuration,
    maxRetries: Int,
    power: Double
  ) extends RetryPolicy with RetryPolicyWithBaseDelay with RetryPolicyWithMaxRetries

  final case class DecorrelatedJitter(
    base: FiniteDuration,
    maxRetries: Int
  ) extends RetryPolicy with RetryPolicyWithBaseDelay with RetryPolicyWithMaxRetries

  final case class FullJitter(
    base: FiniteDuration,
    maxRetries: Int
  ) extends RetryPolicy with RetryPolicyWithBaseDelay with RetryPolicyWithMaxRetries

File: modules/admin/ui/src/domain/traversal/value-objects/retry-policy.ts
/**
 * RetryPolicy - How to retry failed operations
 * Mirrors: com.monadial.waygrid.common.domain.model.routing.RetryPolicy
 */
export type RetryPolicy =
  | { readonly type: 'NoRetry' }
  | { readonly type: 'Linear'; readonly delay: string; readonly maxRetries: number }
  | { readonly type: 'Exponential'; readonly baseDelay: string; readonly maxRetries: number }
  | {
      readonly type: 'BoundedExponential'
      readonly baseDelay: string
      readonly maxDelay: string
      readonly maxRetries: number
    }

export const RetryPolicies = {
  NoRetry: (): RetryPolicy => ({ type: 'NoRetry' }),
  Linear: (delay: string, maxRetries: number): RetryPolicy => ({
    type: 'Linear',
    delay,
    maxRetries,
  }),
  Exponential: (baseDelay: string, maxRetries: number): RetryPolicy => ({
    type: 'Exponential',
    baseDelay,
    maxRetries,
  }),
  BoundedExponential: (
    baseDelay: string,
    maxDelay: string,
    maxRetries: number
  ): RetryPolicy => ({
    type: 'BoundedExponential',
    baseDelay,
    maxDelay,
    maxRetries,
  }),
}

export function formatRetryPolicy(policy: RetryPolicy): string {
  switch (policy.type) {
    case 'NoRetry':
      return 'No retry'
    case 'Linear':
      return `Linear: ${policy.delay}, max ${policy.maxRetries}`
    case 'Exponential':
      return `Exponential: ${policy.baseDelay}, max ${policy.maxRetries}`
    case 'BoundedExponential':
      return `Bounded exp: ${policy.baseDelay}-${policy.maxDelay}, max ${policy.maxRetries}`
  }
}
